### DISCUSSION TITLE
inconsistency when using target object as input to quarto chunk options

### DISCUSSION THREAD

**Discussion by dvictori** [ropensci/tarchetypes#137] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I'm using [targets](https://books.ropensci.org/targets/) to develop a pipeline. At the end of the pipeline I want to generate a report (Quarto markdown) in which some chunks get evaluated depending on objects created during the pipeline.

I can recover the pipeline object using:

```
id <- tar_load(id)
```

and use the variable inside a chunk option:

```
#| echo !expr id
# do some stuff if id == TRUE
```

But, the eval option works differently.

```
#| eval: !expr id

# This should only get evaluated if id is TRUE
# but also raises an error when issuing tar_make()
# Error in eval(x, envir = envir) : object 'choice' not found
1 + 1 == 2
```

The chunk above gets evaluated correctly, that is, if result from expression is TRUE, it evaluates. But it also raises an error when issuing tar_make()
```
Error in eval(x, envir = envir) : object 'choice' not found
```

If I change the code chunk to `#| eval: !expr tar_read(choice)`, than no error is reported

Is this the expected behavior? Or is this an inconsistency?

**Comment by wlandau** [comment]
Hard to say, I would need a reprex as explained at https://books.ropensci.org/targets/help.html.

> I can recover the pipeline object using:

`id <- tar_load(id)`  is problematic since `tar_load()` returns `NULL`. I suggest `id <- tar_read(id)` or just `tar_load(id)`.

**Reply by dvictori** [reply]
Thanks for your reply Will. I tried making a reprex and placed it in this repo

https://github.com/dvictori/target_err_eval

I usually did short reprexes that could be pasted into the tread. But this is the first time I make one with two files. So I'm not sure there is a better way.

**Reply by wlandau** [reply]
Thanks for the reprex. I ran your code and saw the same errors, and your pipeline looks correct.

I got the same output when I removed `targets` from the report. Instead of `tar_load(choice)` or `choice <- tar_read(choice)`, you can see the same errors if you set `choice <- TRUE` and do not invoke `targets` at all. So as to whether some chunks should or should not run, I think this is a Quarto issue and not a `targets`/`tarchetypes` issue. You can ask about that part on https://community.rstudio.com, Stack Overflow, or similar forum.

As for "Error in eval(x, envir = envir) : object 'choice' not found", I get those errors when I run `knitr::knit("report.qmd", tangle = TRUE)` on a smaller version of that report:

    ---
    format: html
    ---
    
    ```{r}
    #| eval: !expr choice
    NULL
    ```
    
    ```{r}
    #| eval: !expr (!choice)
    NULL
    ```

That's another good question for a general forum (why this Quarto report cannot tangle/purl). `tarchetypes` relies on `knitr::knit(tangle = TRUE)` to find `tar_load()`/`tar_read()` dependencies.

**Reply by dvictori** [reply]
Thanks. I never thought of removing targets. So it looks like it's a quarto issue. I'll try to come up with a reprex and post on stack overflow

**Reply by dvictori** [reply]
posted a question on rstudio community

https://community.rstudio.com/t/error-in-knit-tangle-true-when-using-eval-chunk-options/158358

Hope I described everything right

**Reply by dvictori** [reply]
An answer came. Not sure I understood it all. But it looks like variables can't be used on chunk headers. That's why using the variable raises an error but using `tar_read(choice)` does not.

https://community.rstudio.com/t/error-in-knit-tangle-true-when-using-eval-chunk-options/158358/5

And the answer links to an knitr issue that you already commented. 

https://github.com/yihui/knitr/issues/1938#issuecomment-766871863

So I guess is that not much can be done
